### DISCUSSION TITLE
Sparse index in Postgresql

### DISCUSSION THREAD

**Discussion by HonzaKopecky** [prisma/prisma#24330] [Q&A] [answered]
### Bug description

I am trying to create a sparse index on my schema. I know this is not supported by prisma just yet but I am trying to achieve this through a manual migration. The problem is that after applying my updated migration file using `npx prisma migrate dev`, it is applied successfully but another migration is created right away. So I am being asked whether I want to create a new migration (with the same contents as the one created using `--create-only`). It seems to me that prisma is not able to link the schema to the database. Is that possible?

I also tried the alternative approach with just submitting the autogenerated migration, changing it, deleting `_prisma_migrations` entry and running `npx prisma migrate resolve --applied` but that didn't work either.

I am aware I am running an outdated version of Prisma but updating is currently a bit challenging given the project status.

### How to reproduce

1. Add `@@unique([externalId])` at the end of `Company`.
1. Next, I ran `npx prisma migrate dev --create-only`. The following migration file was created:
```
/*
  Warnings:

  - A unique constraint covering the columns `[externalId]` on the table `Company` will be added. If there are existing duplicate values, this will fail.

*/
-- CreateIndex
CREATE UNIQUE INDEX "Company_externalId_key" ON "Company"("externalId");
```
1. Then I updated the mirgration file to:
```
-- CreateIndex
CREATE UNIQUE INDEX "Company_externalId_key" ON "Company"("externalId") WHERE "externalId" IS NOT NULL;
```
1. Then I run `npx prisma migrate dev` to apply the migration. 

### Expected behavior

I would expect that after applying the updated migration successfuly, I would have "clean slate" and could go on.

### Prisma information

```prisma
model Company {
  id                           String     @id @default(uuid())
  externalId                   String?
  source                       String?
  name                         String
  jobOffers                    JobOffer[] @relation("companyInfo")
  description                  String     @default("")
  maxValOfWork                 String     @default("")
  amountOfTandems              String     @default("")
  specialAgreements            String     @default("")
  xmlUrl                       String     @default("")
  logo                         String     @default("")
  emails                       String[] 
  totalJobOffers               Int        @default(0)
  totalTandemJobOffers         Int        @default(0)
  totalArchivedTandemJobOffers Int        @default(0)
  createdAt                    DateTime   @default(now())
  updatedAt                    DateTime   @updatedAt
  sendExtraFile                Boolean    @default(false)
  sendByEmail                  Boolean    @default(false)
}
```

### Environment & setup

- OS: Ubuntu
- Database: PostgreSQL 14.5
- Node.js version: 20.10

### Prisma Version

```
prisma                  : 3.15.2
@prisma/client          : 3.15.2
Current platform        : debian-openssl-3.0.x
Query Engine (Node-API) : libquery-engine 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Migration Engine        : migration-engine-cli 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/migration-engine-debian-openssl-3.0.x)
Introspection Engine    : introspection-core 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/introspection-engine-debian-openssl-3.0.x)
Format Binary           : prisma-fmt 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-3.0.x)
Default Engines Hash    : 461d6a05159055555eb7dfb337c9fb271cbd4d7e
Studio                  : 0.462.0
```

**Comment by charanjit-singh** [comment]
https://github.com/prisma/prisma/issues/3419 as well

**Comment by charanjit-singh** [comment]
Related to https://github.com/prisma/prisma/issues/3076 ?

**Comment by HonzaKopecky** [comment]
It is definitely related but I lived under the impression that those threads are about direct support of the sparse indices. What I am trying to achieve is to implement it using manual migration. But maybe my understanding is wrong here. Is the reason for the above described problem that prisma can't find the actual index in the database and that's why it keeps creating a new migration?

**Reply by janpio** [reply, marked as answer]
Exactly. 

Unfortunately, that is an artifact of Prisma not fully supporting this feature, which leads to Prisma thinking that `@@unique([externalId])` might be missing and creating the migration to create it.

If you did not add the `@@unique` to your schema, just created an empty migration and added that sparse index SQL there, it would _not_ create this SQL to create a normal unique index - but potentially Prisma might try to _drop_ the sparse index that you created manually (as it might be confused why this constraint exists without a definition in Prisma Schema). If that is the case, you will have to remove that SQL to drop the index any time you use `migrate dev` to create a migration in the future.

This will only be properly solved when the feature is fully implemented.
